File: vben-admin-backend/java-backend/src/main/java/com/vben/backend/module/system/controller/SystemMenuController.java
package com.vben.backend.module.system.controller;

import cn.dev33.satoken.annotation.SaCheckRole;
import cn.dev33.satoken.annotation.SaMode;
import com.vben.backend.common.result.R;
import com.vben.backend.module.system.dto.MenuSaveRequest;
import com.vben.backend.module.system.service.SysMenuService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

/**
 * 菜单管理接口（/system/menu/**，仅 super/admin 可访问）。
 *
 * @author Starry
 */
@RestController
@RequestMapping("/system/menu")
@RequiredArgsConstructor
@SaCheckRole(value = {"super", "admin"}, mode = SaMode.OR)
public class SystemMenuController {

    private final SysMenuService menuService;

    /** GET /system/menu/list：菜单扁平列表 */
    @GetMapping("/list")
    public R<List<?>> list() {
        return R.ok(menuService.listAll());
    }

    /** GET /system/menu/tree：菜单树（权限分配勾选用） */
    @GetMapping("/tree")
    public R<List<Map<String, Object>>> tree() {
        return R.ok(menuService.tree());
    }

    /** POST /system/menu：新建菜单 */
    @PostMapping
    public R<Long> create(@RequestBody MenuSaveRequest req) {
        return R.ok(menuService.create(req));
    }

    /** PUT /system/menu/{id}：更新菜单（id 走路径，body 无 id 时以路径为准） */
    @PutMapping("/{id}")
    public R<Void> update(@PathVariable Long id, @RequestBody MenuSaveRequest req) {
        if (req.getId() == null) {
            req.setId(id);
        }
        menuService.update(req);
        return R.ok();
    }

    /** DELETE /system/menu/{id}：删除菜单 */
    @DeleteMapping("/{id}")
    public R<Void> delete(@PathVariable Long id) {
        menuService.remove(id);
        return R.ok();
    }

    /** GET /system/menu/name-exists：菜单名是否重复（排除 id） */
    @GetMapping("/name-exists")
    public R<Boolean> nameExists(@RequestParam String name,
                                 @RequestParam(required = false) Long id) {
        return R.ok(menuService.nameExists(name, id));
    }

    /** GET /system/menu/path-exists：菜单路径是否重复（排除 id） */
    @GetMapping("/path-exists")
    public R<Boolean> pathExists(@RequestParam String path,
                                 @RequestParam(required = false) Long id) {
        return R.ok(menuService.pathExists(path, id));
    }
}

File: vben-admin-backend/node-backend/src/modules/system/system-menu/system-menu.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Query,
  Param,
  UseGuards,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth, ApiQuery } from '@nestjs/swagger';
import { SystemMenuService } from './system-menu.service';
import { MenuSaveDto } from './dto/system-menu.dto';
import { R } from '../../../common/result';
import { JwtAuthGuard } from '../../../common/guards/jwt-auth.guard';
import { PermissionGuard } from '../../../common/guards/permission.guard';
import { Roles } from '../../../common/decorators/roles.decorator';

@ApiTags('系统-菜单管理')
@ApiBearerAuth('JWT')
@Controller('system/menu')
@UseGuards(JwtAuthGuard, PermissionGuard)
@Roles('super', 'admin')
export class SystemMenuController {
  constructor(private readonly menuService: SystemMenuService) {}

  /** GET /system/menu/list：菜单扁平列表 */
  @Get('list')
  @ApiOperation({ summary: '菜单扁平列表' })
  async list() {
    return R.ok(await this.menuService.listAll());
  }

  /** GET /system/menu/tree：菜单树 */
  @Get('tree')
  @ApiOperation({ summary: '菜单树（权限分配用）' })
  async tree() {
    return R.ok(await this.menuService.tree());
  }

  /** POST /system/menu：新建菜单 */
  @Post()
  @ApiOperation({ summary: '新建菜单' })
  async create(@Body() dto: MenuSaveDto) {
    const id = await this.menuService.create(dto);
    return R.ok(id);
  }

  /** PUT /system/menu/:id：更新菜单 */
  @Put(':id')
  @ApiOperation({ summary: '更新菜单' })
  async update(@Param('id') id: string, @Body() dto: MenuSaveDto) {
    await this.menuService.update(id, dto);
    return R.ok();
  }

  /** DELETE /system/menu/:id：删除菜单 */
  @Delete(':id')
  @ApiOperation({ summary: '删除菜单' })
  async delete(@Param('id') id: string) {
    await this.menuService.remove(id);
    return R.ok();
  }

  /** GET /system/menu/name-exists：菜单名是否重复 */
  @Get('name-exists')
  @ApiOperation({ summary: '检查菜单名是否已存在' })
  @ApiQuery({ name: 'name', description: '菜单名' })
  @ApiQuery({ name: 'id', required: false, description: '排除的 ID（编辑时用）' })
  async nameExists(
    @Query('name') name: string,
    @Query('id') id?: string,
  ) {
    return R.ok(await this.menuService.nameExists(name, id));
  }

  /** GET /system/menu/path-exists：菜单路径是否重复 */
  @Get('path-exists')
  @ApiOperation({ summary: '检查菜单路径是否已存在' })
  @ApiQuery({ name: 'path', description: '路由路径' })
  @ApiQuery({ name: 'id', required: false, description: '排除的 ID（编辑时用）' })
  async pathExists(
    @Query('path') path: string,
    @Query('id') id?: string,
  ) {
    return R.ok(await this.menuService.pathExists(path, id));
  }
}
